### PR TITLE
[WIP] Prevented unnecessary table creation in tests.

### PR DESCRIPTION
**Pull request by charettes**
83 tables so far and counting (excluding automatically created intermediary M2M).

This should speed up the test suite startup on backends with slow table creation and cloning (Oracle, MSSQL) and also help with the MSSQL constraint toggling issue.

### CODE REVIEW

**Comment by charettes** [comment]
I managed to reduce the total number of initially created test tables (excluding the ones created by migrations or schema tests) form 1293 to 1142 (~10%).

I also identified many occurrence of inlined model definition that didn't use an isolated registry resulting in global apps and model._meta cache busting. An other issue is that since these models are registered to the global apps cache and following call to `migrate` will actually create their associated table.

I'd like to introduce two helpers to correctly fix this issue:
1. A `models.Model` subclass or decorator that mark models solely used for their definition in order to differentiate them from the ones that are explicitly unmanaged for specific purpose.
2. A test decorator and context manager that makes models defined in it's context/scope register to an isolated registry. This would avoid all the boiler plate actually required to achieve model definition isolation.

**Comment by charettes** [comment]
Buildbot, test on Oracle.

**Comment by charettes** [comment]
Buildbot, test on selenium.

**Comment by carljm** [comment]
@charettes Great work! And your plan makes sense to me. (1) addresses the main concern I had with the first draft of this PR (ambiguity about why a test model was un-managed).

File: tests/foreign_object/tests.py [hunk 1]
```diff
@@ -431,16 +435,22 @@ class Child(models.Model):
                 related_name='children',
             )
 
+            class Meta:
+                apps = apps = test_apps
```
Comment (poleha) [inline tests/foreign_object/tests.py]: typo?

Comment (charettes) [inline tests/foreign_object/tests.py]: Indeed!

**Comment by timgraham** [comment]
Simon, please reopen if you return to this.

**Pull request closed without merging**